ecs-agent/manageddaemon/managed_daemon.go

```diff
@@ -0,0 +1,326 @@
+	// identified by the volume name `agentCommunicationMount`
+	// the SourceVolumeHostPath will always be overridden to
+	// /var/run/ecs/<md.imageName>
+	agentCommunicationMount *MountPoint
+
+	// Daemons require an application log mount
+	// identified by the volume name `applicationLogMount`
+	// the SourceVolumeHostPath will always be overridden to
+	// /var/log/ecs/<md.imageName>
+	applicationLogMount *MountPoint
+
+	mountPoints []*MountPoint
+	environment map[string]string
+
+	loadedDaemonImageRef string
+}
+
+// A valid managed daemon will require
+// healthcheck and mount points to be added
+func NewManagedDaemon(
+	imageName string,
+	imageRef string,
+	imageTag string,
+) *ManagedDaemon {
+	if imageTag == "" {
+		imageTag = imageTagDefault
+	}
+	// health check retries 0 is valid
+	// we'll override this default to -1
+	newManagedDaemon := &ManagedDaemon{
+		imageName:          imageName,
+		imageRef:           imageRef,
+		imageTag:           imageTag,
+		healthCheckRetries: -1,
+	}
+	return newManagedDaemon
+}
+
+// ImportAll function will parse/validate all managed daemon definitions
+// defined in /var/lib/ecs/deps/daemons and will return an array
+// of valid ManagedDeamon objects
+func ImportAll() []*ManagedDaemon {
+	// TODO parse taskdef json files in /deps/daemons
+	// TODO validate that each daemon has a corresponding image tar
+	ebsManagedDaemon := NewManagedDaemon("ebs-csi-driver",
+		"public.ecr.aws/ebs-csi-driver/aws-ebs-csi-driver",
+		"v1.20.0")
+	// TODO add healthcheck
+	// TODO add mount points
+	return []*ManagedDaemon{ebsManagedDaemon}
+}
+
+func (md *ManagedDaemon) SetHealthCheck(
+	healthCheckTest []string,
+	healthCheckInterval time.Duration,
+	healthCheckTimeout time.Duration,
+	healthCheckRetries int) {
+	md.healthCheckInterval = healthCheckInterval
+	md.healthCheckTimeout = healthCheckTimeout
+	md.healthCheckRetries = healthCheckRetries
+	md.healthCheckTest = make([]string, len(healthCheckTest))
+	copy(md.healthCheckTest, healthCheckTest)
+}
+
+func (md *ManagedDaemon) GetImageName() string {
+	return md.imageName
+}
+
+func (md *ManagedDaemon) GetImageRef() string {
+	return md.imageRef
+}
+
+func (md *ManagedDaemon) GetImageTag() string {
+	return md.imageTag
+}
+
+func (md *ManagedDaemon) GetImageCanonicalRef() string {
+	return (fmt.Sprintf("%s:%s", md.imageRef, md.imageTag))
+}
+
+func (md *ManagedDaemon) GetImageTarPath() string {
+	return (fmt.Sprintf("%s/%s", imageTarPath, md.imageName))
+}
+
+func (md *ManagedDaemon) GetAgentCommunicationMount() *MountPoint {
+	return md.agentCommunicationMount
+}
+
+func (md *ManagedDaemon) GetApplicationLogMount() *MountPoint {
+	return md.applicationLogMount
+}
+
+// returns list of mountpoints without the
+// agentCommunicationMount and applicationLogMount
+func (md *ManagedDaemon) GetFilteredMountPoints() []*MountPoint {
+	filteredMounts := make([]*MountPoint, len(md.mountPoints))
+	copy(filteredMounts, md.mountPoints)
```

Comment:
as I understand, md.mountPoints does not include `agentCommunicationMount` and `applicationLogMount`. if this is true then we could do `return md.mountPoints` without doing a copy.

Reply:
I like to be sure that getters return copies when there are references (ie *MountPoint) so I don't end up accidentally modifying and can only do so using my setter.